### ISSUE TITLE
Deprecate `BasePlotter.mesh`, `BasePlotter.mapper`, and `BasePlotter.volume`

### ISSUE THREAD

**Issue by banesullivan** [pyvista/pyvista#3806]
We assign `BasePlotter.mesh` and `BasePlotter.mapper` during plotting due to some legacy design choices for the methods `update_scalars()`, `update_coordinates()`, and `add_scalar_bar()`. Removing these assignments will clean up some of the code in `add_mesh()` and `add_volume()` as well as improve GC on the `BasePlotter`. For instance we have to make sure these are cleaned up here:

https://github.com/pyvista/pyvista/blob/61d16855746ad5aa9d63fd2e531bc8b880558cb0/pyvista/plotting/plotting.py#L3870-L3871

and here:

https://github.com/pyvista/pyvista/blob/61d16855746ad5aa9d63fd2e531bc8b880558cb0/pyvista/plotting/plotting.py#L4134-L4135

and here:

https://github.com/pyvista/pyvista/blob/61d16855746ad5aa9d63fd2e531bc8b880558cb0/pyvista/plotting/plotting.py#L4175-L4177

Specifically, `update_scalars()` and `update_coordinates()` are for updating a mesh that was previously added to the scene. IMO, these methods should not be on the Plotter, and any change to the mesh object should be registered by the rendering pipeline (already possible but improved by #3318). These methods allow for their `mesh` argument to be `None`, defaulting to the last added mesh (`self.mesh`):

https://github.com/pyvista/pyvista/blob/61d16855746ad5aa9d63fd2e531bc8b880558cb0/pyvista/plotting/plotting.py#L4031=L4032

https://github.com/pyvista/pyvista/blob/61d16855746ad5aa9d63fd2e531bc8b880558cb0/pyvista/plotting/plotting.py#L4089-L4090

On the other hand, `self.mapper` is only used by `add_scalar_bar()` in a specific scenario:

https://github.com/pyvista/pyvista/blob/61d16855746ad5aa9d63fd2e531bc8b880558cb0/pyvista/plotting/plotting.py#L3989-L3994

which may be easier to handle/deprecate.

## Proposed Deprecation to clean up `self.mesh`

I'd like to propose a deprecation of the `mesh=None` argument to `update_scalars()` and`update_coordinates()`, removing these methods' reliance on `self.mesh`. Further I'd like to propose a full deprecation of `update_coordinates()` as updating the coordinates of a mesh should propagate to the plotter regardless. The only additional benefit `update_coordinates()` has is the render call which is marginal as the LoC to use `update_coordinates()` is often more than updating the mesh in-place and calling `render()`.

### Demonstration of existing solution

You can demonstrate this yourself by changing `examples/02-plot/gif.py` as follows:

```diff
diff --git a/examples/02-plot/gif.py b/examples/02-plot/gif.py
index 1c6d1b3b..2c611b49 100644
--- a/examples/02-plot/gif.py
+++ b/examples/02-plot/gif.py
@@ -44,8 +44,7 @@ pts = grid.points.copy()
 nframe = 15
 for phase in np.linspace(0, 2 * np.pi, nframe + 1)[:nframe]:
     z = np.sin(r + phase)
-    pts[:, -1] = z.ravel()
-    plotter.update_coordinates(pts, render=False)
+    grid.points[:, -1] = z.ravel()
     plotter.update_scalars(z.ravel(), render=False)

     # Write a frame. This triggers a render.
```

Further, the same is true for `update_scalars()`:

```diff
diff --git a/examples/02-plot/gif.py b/examples/02-plot/gif.py
index 1c6d1b3b..1fed16d1 100644
--- a/examples/02-plot/gif.py
+++ b/examples/02-plot/gif.py
@@ -23,15 +23,15 @@ z = np.sin(r)

 # Create and structured surface
 grid = pv.StructuredGrid(x, y, z)
+grid['Height'] = z.ravel()

 # Create a plotter object and set the scalars to the Z height
 plotter = pv.Plotter(notebook=False, off_screen=True)
 plotter.add_mesh(
     grid,
-    scalars=z.ravel(),
+    scalars='Height',
     lighting=False,
     show_edges=True,
-    scalar_bar_args={"title": "Height"},
     clim=[-1, 1],
 )

@@ -46,7 +46,7 @@ for phase in np.linspace(0, 2 * np.pi, nframe + 1)[:nframe]:
     z = np.sin(r + phase)
     pts[:, -1] = z.ravel()
     plotter.update_coordinates(pts, render=False)
-    plotter.update_scalars(z.ravel(), render=False)
+    grid['Height'] = z.ravel()

     # Write a frame. This triggers a render.
     plotter.write_frame()
```

### Cost/Benefit of Deprecation

I have a feeling that `update_coordinates()` will go unmissed but deprecating `update_scalars()` may stir up some frustration as it has been more prevalently used by those who pass `ndarray`s as the `scalars` argument to `add_mesh()` (which IMO should also be deprecated but that's a topic for another day).

`update_scalars()` and `update_coordinates()` are the only things dependent on the legacy assignment of `self.mesh` in `BasePlotter`. So I see a clear way to clean up `add_mesh()`, reduce code complexity, improving GC, and move towards a clear, well-defined API for updating the coordinates and scalars of a mesh (by modifying the mesh directly and not going through the plotter)

## Proposed Deprecation to clean up `self.mapper`

This should be a bit easier as I don't think this method is used all that much in the public API. I think adding a deprecation warning for `mapper=None` will go pretty smoothly and then users will just have to use some sort of logic like the following now that `Actor` and `Mapper` are wrapped in PyVista:

```py
pl = pv.Plotter()
actor = pl.add_mesh(mesh, show_scalar_bar=False)
pl.add_scalar_bar("title", mapper=actor.mapper)
```

## Proposed Deprecation to clean up `self.volume`

AFAIK, `self.volume` isn't actually used and was implemented to match the convention of `add_mesh()` so a quick removal should go unnoticed

**Comment by banesullivan** [comment]
With this work, we probably want to clean up `self.textActor` too:

https://github.com/pyvista/pyvista/blob/61d16855746ad5aa9d63fd2e531bc8b880558cb0/pyvista/plotting/plotting.py#L4178

**Comment by akaszynski** [comment]
I agree with this depreciation, but I think we can go a step further to ensure backwards compatibility.

`Plotter.renderer.actors` contains references to all added `pyvista.DataSet` objects via the `actor.mapper.dataset` attribute. All we need to do is access the last added mesh (the active mesh) and use that.

### Proposal

```py
import pyvista as pv

mesh_a = pv.Sphere()
mesh_b = pv.Sphere()
mesh_c = pv.Sphere()

pl = pv.Plotter()
pl.add_mesh(mesh_a)
pl.add_mesh(mesh_b)
pl.add_mesh(mesh_c)

def active_mesh(pl):
    actor = list(pl.renderer.actors.values())[-1]
    return actor.mapper.dataset

assert active_mesh(pl) is mesh_c
```

This works because `dict` is ordered by default in Python 3.7 and newer. We can make this explicit by using [OrderedDict](https://docs.python.org/3/library/collections.html#collections.OrderedDict).

This way we'll just grab the last added mesh and we can deprecate the `mesh` attribute without changing the behavior.

**Comment by akaszynski** [comment]
If we want to go crazy, we can make `mesh` a read only attribute of the `Plotter`, but I would prefer deprecating it fully and instead keeping it private (either property or as a getter function) and using the new `actor` and `mapper` APIs.

**Comment by banesullivan** [comment]
Aha! In this work, I'll could add an `active_dataset` property perhaps that will work for both `add_mesh` and `add_volume`? Instead of just `mesh`

**Comment by akaszynski** [comment]
That's perfect. Add in a deprecation for the `mesh` attribute that sticks around for a release or two.

**Comment by banesullivan** [comment]
cc @germa89, I just noticed your #4779 which may have been at odds with this effort

**Comment by germa89** [comment]
> cc @germa89, I just noticed your #4779 which may have been at odds with this effort

History of my life ... sigh.........

Well. #4779 was made with the intention to make accessible the meshes from the plotter. I think this is very useful for [PyMAPDL](https://github.com/ansys/pymapdl) since we compute the meshes behind the scenes, and as much we just return the plotter object.
Until that PR, Pyvista only let you access the last one using `BasePlotter.mesh` which I think it wasn't coherent if you had more meshes.

Regarding the current PR, I think deprecating `BasePlotter.mesh` or `BasePlotter.meshes` might break some stuff... but I'm happy to work on it when it happens (I hope I won't regret saying this 😆 )

**Comment by banesullivan** [comment]
In hindsight, I think your method in #4779 is compatible if we deprecate this as its not adding any references and is more of a "convenience method" to iterate over the actors

To be honest, while I want to advocate for this deprecation strongly, I don't have the time to do it... so it probably won't get done unless someone else with the time feels as strongly as I do
